toolchain/check/handle_loop_statement.cpp

```diff
@@ -40,17 +67,57 @@ auto HandleForStatement(Context& context, Parse::Node parse_node) -> bool {
   return context.TODO(parse_node, "HandleForStatement");
 }
 
-auto HandleWhileCondition(Context& context, Parse::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleWhileCondition");
-}
-
 auto HandleWhileConditionStart(Context& context, Parse::Node parse_node)
     -> bool {
-  return context.TODO(parse_node, "HandleWhileConditionStart");
+  // Branch to the loop entry block.
+  auto loop_entry_id = context.AddDominatedBlockAndBranch(parse_node);
+  context.node_block_stack().Pop();
+
+  // Start emitting the loop entry block.
+  context.node_block_stack().Push(loop_entry_id);
+  context.AddCurrentCodeBlockToFunction();
+
+  context.node_stack().Push(parse_node, loop_entry_id);
+  return true;
+}
+
+auto HandleWhileCondition(Context& context, Parse::Node parse_node) -> bool {
+  auto cond_value_id = context.node_stack().PopExpression();
+  auto loop_entry_id =
+      context.node_stack().Peek<Parse::NodeKind::WhileConditionStart>();
+  cond_value_id = ConvertToBoolValue(context, parse_node, cond_value_id);
+
+  // Branch to either the loop body or the loop exit block.
+  auto loop_body_id =
+      context.AddDominatedBlockAndBranchIf(parse_node, cond_value_id);
+  auto loop_exit_id = context.AddDominatedBlockAndBranch(parse_node);
+  context.node_block_stack().Pop();
+
+  // Start emitting the loop body.
+  context.node_block_stack().Push(loop_body_id);
+  context.AddCurrentCodeBlockToFunction();
+  context.break_continue_stack().push_back(
+      {.break_target = loop_exit_id, .continue_target = loop_entry_id});
```

Comment:
FWIW, `loop_body` might be reasonably called `loop_latch` if you like, but there isn't the confusion here for me so really whatever folks like.

Reply:
Looking at the test case and how things get split subsequently for conditions, this doesn't end up the latch, so maybe body is a better term after all.

Reply:
Yeah, this is first block in the body, not necessarily the block with the backedge.